app/controllers/concerns/saml_idp_auth_concern.rb

```diff
@@ -114,4 +117,53 @@ def current_service_provider
   def current_issuer
     @_issuer ||= saml_request.service_provider.identifier
   end
+
+  def request_url
+    # puts "Original URL: #{request.original_url}"
+    url = URI.parse request.original_url
+    query_params = parse_query_params url.query
+    unless query_params['SAMLRequest']
+      orig_request = saml_request.options[:get_params][:SAMLRequest]
+      query_params['SAMLRequest'] = orig_request
+    end
+
+    url.query = query_hash_to_string(query_params)
+    # puts "Modified URL: #{url.to_s}"
+    url.to_s
+  end
+
+  # :reek:FeatureEnvy
+  # :reek:TooManyStatements
+  # rubocop:disable Metrics/MethodLength
+  # Derived from https://github.com/postmodern/uri-query_params/blob/master/lib/uri/query_params/query_params.rb#L44-L65
+  def parse_query_params(query_string)
```

Comment:
Parsing URLs on our own always makes me a bit nervous. Is there a reason not to use `Rack::Utils.parse_nested_query` here?